bokehjs/src/coffee/models/plots/plot_canvas.coffee

```diff
@@ -448,14 +448,20 @@ class PlotCanvasView extends Renderer.View
         @canvas.get("canvas_height") != height)
       @canvas_view.set_dims([width, height], trigger=false)
 
-    @canvas_view.render(force_canvas)
-
     for k, v of @renderer_views
       if not @range_update_timestamp? or v.set_data_timestamp > @range_update_timestamp
         @update_dataranges()
         break
 
     ctx = @canvas_view.ctx
+    ctx.save()  # Save default state
+
+    # Set hidpi-transform
+    ratio = @canvas_view.render(force_canvas)
```

Comment:
`Canvas.render` always returns ratio, I think you missed an indent :)  but I can see your concern for render returning something. Maybe we should just rename that method, because it does not render anything, it makes sure the canvas is setup correctly.

That said, are you ok if we fix this in the PR that I'll make against @birdsarah 's branch? It requires changes to that method anyway, and I'd like to minimize the code changes to that code here to avoid conflicts.